From 369f7fe871c1f97d876bb75cd2119b8683679076 Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Sat, 10 Oct 2026 05:36:45 +0200
Subject: [PATCH] Polycrystal: errors of the options of the non linear solver

An initial adaptive tolerance given without adaptive tolerance is refused,
and the errors thrown by the constructor of the problem (invalid options
of the non linear solver) are reported before exiting.
---
 polycrystal/UniaxialPolycrystal.cxx | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/polycrystal/UniaxialPolycrystal.cxx b/polycrystal/UniaxialPolycrystal.cxx
index 755b0f6..aeb93c4 100644
--- a/polycrystal/UniaxialPolycrystal.cxx
+++ b/polycrystal/UniaxialPolycrystal.cxx
@@ -7,6 +7,8 @@
  */
 
 #include <memory>
+#include <string>
+#include <exception>
 #include <csignal>
 #include <cstdlib>
 #include <fstream>
@@ -73,7 +75,7 @@ struct NumericalParameters {
    * `EisenstatWalker2` (see `AdaptiveLinearSolverTolerance` in mfem-mgis)
    */
   const char *adaptive_linear_solver_tolerance = "None";
-  //! \brief initial adaptive tolerance, the default one if negative
+  //! \brief initial adaptive tolerance, the default one if not given
   mfem_mgis::real adaptive_linear_solver_initial_tolerance = -1;
 };
 
@@ -185,9 +187,17 @@ int main(int argc, char *argv[]) {
   // options of the non linear solver: adaptive tolerance of its linear
   // solver, if required
   auto non_linear_solver_options = mfem_mgis::Parameters{};
+  const auto initial_tolerance_given =
+      p.adaptive_linear_solver_initial_tolerance != -1;
+  if ((std::string_view{p.adaptive_linear_solver_tolerance} == "None") &&
+      (initial_tolerance_given)) {
+    mfem_mgis::Profiler::Utils::Message(
+        "error: an initial tolerance is given without adaptive tolerance");
+    std::exit(EXIT_FAILURE);
+  }
   if (std::string_view{p.adaptive_linear_solver_tolerance} != "None") {
     auto adaptive_options = mfem_mgis::Parameters{};
-    if (p.adaptive_linear_solver_initial_tolerance > 0) {
+    if (initial_tolerance_given) {
       adaptive_options.insert(ctx, "InitialTolerance",
                               p.adaptive_linear_solver_initial_tolerance) |
           or_die;
-- 
2.43.0

